rolling-update-kubernetes/src/main/scala/akka/rollingupdate/kubernetes/KubernetesApiImpl.scala

```
@@ -0,0 +1,321 @@
+        ))
+    )
+    val httpResponse = makeRequest(
+      httpRequest,
+      s"Timed out updating pod-deletion-cost annotation for pod: [$podName] with cost: [$cost]. Namespace: [$namespace]")
+    httpResponse.map {
+      case HttpResponse(status, _, e, _) if status.isSuccess() =>
+        e.discardBytes()
+        Done
+      case HttpResponse(s @ ClientError(_), _, e, _) =>
+        e.discardBytes()
+        throw new PodCostClientException(s.toString())
+      case HttpResponse(status, _, e, _) =>
+        e.discardBytes()
+        throw new PodCostException(s"Request failed with status=$status")
+    }
+  }
+
+  /*
+  PATH: to get all: /apis/akka.io/v1/namespaces/<namespace>/podcosts
+  PATH: to get a specific one: /apis/akka.io/v1/namespaces/<namespace>/podcosts/<system-name>
+  curl -v -X POST localhost:8080/apis/akka.io/v1/namespaces/<namespace>/podcosts/ -H "Content-Type: application/yaml" --data-binary "@pod-cost-example.yml"
+
+  responds with either:
+  409 Conflict Already Exists
+
+  OR
+
+  201 Created if it works
+   */
+  override def readOrCreatePodCostResource(crName: String): Future[PodCostResource] = {
+    val maxTries = 5
+
+    def loop(tries: Int = 0): Future[PodCostResource] = {
+      log.debug("Trying to create PodCost {}", tries)
+      for {
+        olr <- getPodCostResource(crName)
+        lr <- olr match {
+          case Some(found) =>
+            log.debug("{} already exists. Returning {}", crName, found)
+            Future.successful(found)
+          case None =>
+            log.info("PodCost {} does not exist, creating", crName)
+            createPodCostResource(crName).flatMap {
+              case Some(created) => Future.successful(created)
+              case None =>
+                if (tries < maxTries) loop(tries + 1)
+                else Future.failed(new PodCostException(s"Unable to create or read PodCost after $maxTries tries"))
+            }
+        }
```

Could be the yield block instead and return/throw directly instead of wrapping all return values in successful/failed

---

@johanandren I incorporated some of your suggestions in aa4a6e81, but not sure I understand how to change this to yield and not flatMap. Note that it calls createPodCostResource and loop (recursively) and those return Future

---

I think I simply missed the async-recursive loop.